crates/accelerate/src/synthesis/linear/lnn.rs

```diff
@@ -0,0 +1,217 @@
+
+// Creating the patterns for the phase layers.
+fn _create_patterns(n: isize) -> HashMap<(isize, isize), (isize, isize)> {
+    let (pat1, pat2) = if n % 2 == 0 {
+        (_even_pattern1(n), _even_pattern2(n))
+    } else {
+        (_odd_pattern1(n), _odd_pattern2(n))
+    };
+
+    let ind = if n % 2 == 0 {
+        (2 * n - 4) / 2
+    } else {
+        (2 * n - 4) / 2 - 1
+    };
+
+    HashMap::from_iter((0..n).map(|i| ((0, i), (i, i))).chain(
+        (0..(n / 2)).cartesian_product(0..n).map(|(layer, i)| {
+            (
+                (layer + 1, i),
+                (
+                    pat1[(ind - (2 * layer) + i) as usize],
+                    pat2[((2 * layer) + i) as usize],
+                ),
+            )
+        }),
+    ))
+}
+
+// A single layer of CX gates.
+fn _append_cx_stage1(gates: &mut LnnGatesVec, n: usize) {
+    for i in 0..(n / 2) {
+        gates.push((
+            StandardGate::CXGate,
+            smallvec![],
+            smallvec![Qubit((2 * i) as u32), Qubit((2 * i + 1) as u32)],
+        ))
+    }
+
+    for i in 0..((n + 1) / 2 - 1) {
+        gates.push((
+            StandardGate::CXGate,
+            smallvec![],
+            smallvec![Qubit((2 * i + 2) as u32), Qubit((2 * i + 1) as u32)],
+        ))
+    }
+}
+
+// A single layer of CX gates.
+fn _append_cx_stage2(gates: &mut LnnGatesVec, n: usize) {
+    for i in 0..(n / 2) {
+        gates.push((
+            StandardGate::CXGate,
+            smallvec![],
+            smallvec![Qubit((2 * i + 1) as u32), Qubit((2 * i) as u32)],
+        ))
+    }
+
+    for i in 0..((n + 1) / 2 - 1) {
+        gates.push((
+            StandardGate::CXGate,
+            smallvec![],
+            smallvec![Qubit((2 * i + 1) as u32), Qubit((2 * i + 2) as u32)],
+        ))
+    }
+}
+
+// Appends correct phase gate during CZ synthesis
+fn _append_phase_gate(pat_val: usize, gates: &mut LnnGatesVec, qubit: usize) {
+    // Add phase gates: s, sdg or z
+    if pat_val % 4 == 1 {
+        gates.push((
+            StandardGate::SdgGate,
+            smallvec![],
+            smallvec![Qubit(qubit as u32)],
+        ))
+    } else if pat_val % 4 == 2 {
+        gates.push((
+            StandardGate::ZGate,
+            smallvec![],
+            smallvec![Qubit(qubit as u32)],
+        ))
+    } else if pat_val % 4 == 3 {
+        gates.push((
+            StandardGate::SGate,
+            smallvec![],
+            smallvec![Qubit(qubit as u32)],
+        ))
+    }
+}
+
+// Synthesis of a CZ circuit for linear nearest neighbor (LNN) connectivity,
+// based on Maslov and Roetteler.
+pub(super) fn synth_cz_depth_line_mr(matrix: ArrayView2<bool>) -> (usize, LnnGatesVec) {
+    let num_qubits = matrix.raw_dim()[0];
+    let pats = _create_patterns(num_qubits as isize);
+
+    // s_gates[i] = 0, 1, 2 or 3 for a gate id, sdg, z or s on qubit i respectively
+    let mut s_gates = Array1::<usize>::zeros(num_qubits);
+
+    let mut patlist: Vec<(isize, isize)> = Vec::new();
```

Comment:
Since all numbers are always positive, would it simplify the code by choosing `usize` as type throughout? That would remove the need for all the `as isize` casts below. Or would that make the logic in the pattern creation tedious?

Reply:
Good catch! I got an subtraction with overflow on initial testing but I guess it was unrelated. Should be fixed